### PR TITLE
feat: testoverride ignore/forcefail/forcepass as regexes

### PR DESCRIPTION
**Pull request by M4tteoP**
In this PR:
- `ignore`, `forcefail` and `forcepass` lists are now considered regexes.
- Updated example inside the readme accordingly

Closes https://github.com/fzipi/go-ftw/issues/75.

### CODE REVIEW

**Comment by fzipi** [comment]
I guess you should rebase after moving to a new organization.

File: config/types.go [hunk 1]
```diff
@@ -37,3 +42,10 @@ type FTWTestOverride struct {
 	ForcePass map[string]string `koanf:"forcepass"`
 	ForceFail map[string]string `koanf:"forcefail"`
 }
+
+// FTWTestOverrideRe hold the lists transformed into regexes
+type FTWTestOverrideRe struct {
```
Comment (anuraaga) [inline config/types.go:47]: koanf seems to support `TextUnmarshaler` so I believe you can avoid needing a second field / struct which makes the configuration a bit harder to reason about

Something like

```go

type FTWTestOverride struct {
  Ignore map[string]FTWRegexp
}

type FTWRegexp regexp.Regexp

func (r *FTWRegexp) UnmarshalText(b []byte) error {
  re, err := regexp.Compile(string(b))
  if err != nil {
    return fmt.Errorf("invalid regexp: %w", err)
  }
  *r = re
  return nil
}
```

File: config/types.go [hunk 2]
```diff
@@ -21,8 +25,9 @@ var FTWConfig *FTWConfiguration
 type FTWConfiguration struct {
 	LogFile             string          `koanf:"logfile"`
 	TestOverride        FTWTestOverride `koanf:"testoverride"`
-	LogMarkerHeaderName string          `koanf:"logmarkerheadername"`
```
[the remark this answers is not in the document]
Comment (M4tteoP) [reply config/types.go:24]: Sorry, I just added `TestOverrideRe` in the middle of the struct, so `LogMarkerHeaderName` appears to be removed, but actually is right after `TestOverrideRe`. If I manage to implement anuraaga suggestion, this confusing line will be removed.

[the remark this answers is not in the document]
Comment (theseion) [reply config/types.go:24]: You're right. Sorry :)

File: config/types.go (hunk 1 above)
Comment (M4tteoP) [reply config/types.go]: Implemented, thank you! I had to update Koanf and the resulting map is `map[*FTWRegexp]string`. If we want to keep the excluded id - comment map, the key was the id (now regex) and the string value is the comment.

File: README.md [hunk 3]
```diff
@@ -184,7 +184,8 @@ Tests can be altered using four lists:
 - `forcepass` is for tests you want to pass unconditionally. You should add a comment on why you force to pass the test
 - `forcefail` is for tests you want to fail unconditionally. You should add a comment on why you force to fail the test
 
-Example using all the lists above:
+The match is performed using Go regexp. You can whether match a specific test or all of them about a specific rule, or file.
```
[the remark this answers is not in the document]
Comment (M4tteoP) [reply README.md:187]: Hi! Thanks, I updated the Readme. I think that they are still lists. Each element of the list permits to have 1-n matches against test IDs, but for each ignore/forcefail/forcepass we can have more than one regex. E.g:
Eg:
```
  forcepass:
    '123456-02$': 'This test will always pass'
    '123457-.*': 'All the tests about rule 123457 will always pass' 
```
Waiting for your feedback about the updated proposal :) Thank you!

File: config/types.go (hunk 1 above)
Comment (theseion) [reply config/types.go]: Oh. Yes, you're absolutely right. I don't know what I was thinking...

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (theseion) [reply README.md]: Oh. Yes, you're absolutely right. I don't know what I was thinking...

File: check/base.go [hunk 4]
```diff
@@ -60,20 +62,32 @@ func (c *FTWCheck) SetNoLogContains(contains string) {
 
 // ForcedIgnore check if this id need to be ignored from results
 func (c *FTWCheck) ForcedIgnore(id string) bool {
-	_, ok := c.overrides.Ignore[id]
-	return ok
+	for re, _ := range c.overrides.Ignore {
+		if (*regexp.Regexp)(re).MatchString(id) {
```
Comment (anuraaga) [inline check/base.go:66]: Let's go ahead and add matchstring to the type itself

```
func (r *FTWRegexp) MatchString(s string) bool {
	return (*regexp.Regexp)(r).MatchString(s)
}
```

File: config/config.go [hunk 5]
```diff
@@ -69,7 +72,7 @@ func NewConfigFromEnv() error {
 
 // NewConfigFromString initializes the configuration from a yaml formatted string. Useful for testing.
 func NewConfigFromString(conf string) error {
-	// kaonf merges by default but we never want to merge in this case
+	// kaonf merges by default, but we never want to merge in this case
```
Comment (theseion) [inline config/config.go:75]:
```suggestion
	// koanf merges by default, but we never want to merge in this case
```

File: config/config.go [hunk 6]
```diff
@@ -39,14 +39,17 @@ func NewConfigFromFile(cfgFile string) error {
...
+	}
 	loadDefaults()
 
 	return err
 }
 
 // NewConfigFromEnv reads configuration information from environment variables that start with `FTW_`
 func NewConfigFromEnv() error {
-	// kaonf merges by default but we never want to merge in this case
+	// kaonf merges by default, but we never want to merge in this case
```
Comment (theseion) [inline config/config.go:52]:
```suggestion
	// koanf merges by default, but we never want to merge in this case
```

File: config/config.go [hunk 7]
```diff
@@ -14,7 +14,7 @@ import (
 // NewConfigFromFile reads configuration information from the config file if it exists,
 // or uses `.ftw.yaml` as default file
 func NewConfigFromFile(cfgFile string) error {
-	// kaonf merges by default but we never want to merge in this case
+	// kaonf merges by default, but we never want to merge in this case
```
Comment (theseion) [inline config/config.go:17]:
```suggestion
	// koanf merges by default, but we never want to merge in this case
```

File: config/config_test.go [hunk 8]
```diff
@@ -34,111 +35,112 @@ var jsonConfig = `
...
-	assert.NoError(t, err)
+    err := NewConfigFromString(yamlConfig)
+    assert.NoError(t, err)
 }
 
 func TestNewConfigFromEnv(t *testing.T) {
-	// Set some environment so it gets merged with conf
-	os.Setenv("FTW_LOGFILE", "kaonf")
+    // Set some environment so it gets merged with conf
+    os.Setenv("FTW_LOGFILE", "kaonf")
```
Comment (theseion) [inline config/config_test.go:95]:
```suggestion
    os.Setenv("FTW_LOGFILE", "koanf")
```

File: config/config_test.go [hunk 9]
```diff
@@ -34,111 +35,112 @@ var jsonConfig = `
...
 
-	err := NewConfigFromEnv()
-	assert.NoError(t, err)
+    err := NewConfigFromEnv()
+    assert.NoError(t, err)
 
-	assert.Equal(t, "kaonf", FTWConfig.LogFile)
+    assert.Equal(t, "kaonf", FTWConfig.LogFile)
```
Comment (theseion) [inline config/config_test.go:100]:
```suggestion
    assert.Equal(t, "koanf", FTWConfig.LogFile)
```

File: config/types.go [hunk 10]
```diff
@@ -32,8 +37,24 @@ type FTWConfiguration struct {
...
+	Input     test.Input            `koanf:"input"`
+	Ignore    map[*FTWRegexp]string `koanf:"ignore"`
+	ForcePass map[*FTWRegexp]string `koanf:"forcepass"`
+	ForceFail map[*FTWRegexp]string `koanf:"forcefail"`
+}
+
+type FTWRegexp regexp.Regexp
+
+func (r *FTWRegexp) UnmarshalText(b []byte) error {
+	fmt.Println("Unmarshalling!")
```
Comment (theseion) [inline config/types.go:49]: Use the logger. I don't think we want to see this message all the time. I also think the text could be more helpful. Maybe just remove the line entirely.

**Review by theseion — CHANGES_REQUESTED** [review]

File: config/types.go (hunk 10 above)
Comment (M4tteoP) [reply config/types.go]: Thanks, my bad, just a debugging print, not intended to be here.

**Review by theseion — APPROVED** [review]

**Review by anuraaga — APPROVED** [review]

**Pull request merged**